lib/open_food_network/enterprise_injection_data.rb

```diff
@@ -1,28 +1,51 @@
 module OpenFoodNetwork
   class EnterpriseInjectionData
     def active_distributor_ids
-      @active_distributor_ids ||=
-        Enterprise.distributors_with_active_order_cycles.ready_for_checkout.pluck(:id)
+      @active_distributor_ids ||= begin
+        CacheService.cached_data_by_class("active_distributor_ids", Enterprise) do
+          Enterprise.distributors_with_active_order_cycles.ready_for_checkout.pluck(:id)
+        end
+      end
     end
 
     def earliest_closing_times
-      @earliest_closing_times ||= OrderCycle.earliest_closing_times
+      @earliest_closing_times ||= begin
+        CacheService.cached_data_by_class("earliest_closing_times", OrderCycle) do
+          OrderCycle.earliest_closing_times
+        end
+      end
     end
 
     def shipping_method_services
-      @shipping_method_services ||= Spree::ShippingMethod.services
+      @shipping_method_services ||= begin
+        CacheService.cached_data_by_class("shipping_method_services", Spree::ShippingMethod) do
+          Spree::ShippingMethod.services
```

Comment:
@luisramos0 You don't like this bit?

Here for example when we call `Spree::ShippingMethod#services` we're running a huge query which involves a join across all shipping methods and all distributors in the entire database, which is really nasty. When the query is done we parse _each item_ via `#map` to arrange it in a giant hash in a specific way. We call this method really often! :see_no_evil: 

With `CacheService#cached_data_by_class` we can cache the resulting Hash and keep serving it (in about a millisecond) until the next time somebody changes a shipping method, at which point the cache entry will be automatically expired.

What's not to love? :sweat_smile:

Reply:
And I am recognizing the marvelous feelings of the start of a caching project and also thinking about the very painful ends of such projects (remember the end of product cache?).

Are you now more convinced this is the way to go? I am really worried about the complexity introduced and all the trouble this can cause #twohardthings

how is the cache expired "automatically"?

Reply:
I share your feelings @luisramos0 . If we go down this path we need be extra careful but pretty please, @Matt-Yorkley let's split this up and go bit by bit so we can measure.

As for the cache expiration @luisramos0 I believe https://signalvnoise.com/posts/3113-how-key-based-cache-expiration-works is the established reference, in the Rails community at least.

Reply:
thanks for the link Pau! I now understand the cached_data_by_class code.

It still depend on the "touch" calls. Are we sure the touch calls introduced in this PR are enough for all the cached data? I'd say we can only do this if we cover each data update case with unit tests. For example, in the case of active_distributor_ids, we need to be sure that when an OC is added to an enterprise this cache key is invalidated (the Enterprise object wont be updated), right? Is that happening?

Reply:
> how is the cache expired "automatically"?

I'll try to clarify it a bit here for reference...

Lets look at an example of automatically expiring cache keys, where we fetch all products in the database but we cache the content per day:

```ruby
day = Time.now.strftime("%Y-%m-%d") # => "2020-04-14"

Rails.cache.fetch("auto-expire-after-#{day}") do
   Products.all
end
```

When we pass a key to `Rails.cache.fetch`, it will return the cached result if a cache entry with that key exists, otherwise run whatever is in the block and save it under the provided key.

So the first time the above code runs on a given day, the result will be cached under a cache key called for example `auto-expire-after-2020-04-14`. For the rest of the day, any time this code runs, the key we're _passing in_ to `#fetch` will still be `auto-expire-after-2020-04-14`, so Rails will return the cached value instead of running the query, because an entry with the specified key exists. 

As soon as this code runs after midnight, the key being passed in will _change_ to: `auto-expire-after-2020-04-15`. At this point a key with that value is _not found_ in the cache, so Rails runs the query in the block (`Products.all`), and saves the result under the specified key. Then for the rest of the day this code will serve the new cached value and not run the query, and so on. The cache entry automatically expires when `Time.now.strftime("%Y-%m-%d")` changes.

Side note; luckily Memcached automatically handles removal of old / unused cache entries so we don't have to worry about cleaning up. that's one of the main things it is designed to do, and do well.

So now another example...

```ruby
most_recently_updated_product_in_the_database = Product.maximum(:updated_at)
timestamp = most_recently_updated_product_in_the_database.to_i

Rails.cache.fetch("auto-expire-me-after-#{timestamp}") do
   Products.all
end
```

In this example we get the `:updated_at` value of the most recently updated `Product` record in the database. We make a timestamp from that value, eg: `1586822400`. We then cache the big product query, using the timestamp in the key. Something like: `auto-expire-me-after-1586822400`.

The first time this code runs, the corresponding key will not be found, so Rails will run the query and save the result under the given cache key. If we run it again and again it will serve the cached content and _not_ trigger the query.

If we then _update and save_ a `Product` record (any product), the timestamp we're generating in the key will not match the old one, the cache key will _not_ be found, Rails will run the query and will save the new result under the new key (eg: `auto-expire-me-after-1586999999`). So: the cache entry in this code is auto-expired whenever a `Product` is updated in the database.

Reply:
> we need to be sure that when an OC is added to an enterprise this cache key is invalidated

When an `Exchange` is updated, we already `touch` the receiver (distributor). We definitely need some extra specs!

Reply:
great, thanks for the details, that's what I understood from the link above :+1:
To complete your detailed description, in that product cache we are including variants of the product. If a variant is updated, the product must be touched, sometimes this happens implicitly (some callback), sometimes it happens explicitly (touch).
The problem with this approach is really object composition, you need to make sure that any change to any part of the product serialization will touch the product object.

The checklist from any entry like this would be:
A - what objects compose the serialization of the main object? this needs to be recursive!! for example, updates to stock items must touch products in some way :see_no_evil: 
B - write a unit test that makes sure an update to every part of the payload will touch the main object (and therefore refresh the cache).

Reply:
Ok, it turns out three of the six queries here ultimately depend on the `OrderCycle.active` scope, so they won't work with normal `touch` calls:

https://github.com/openfoodfoundation/openfoodnetwork/blob/f366aa260516788e9c283ec9785206420a52654a/app/models/order_cycle.rb#L29-L33

I've cut those three.

Reply:
> updates to stock items must touch products in some way

Yes, the example was a simple one. I've intentionally avoided using this on anything to do with actual products because it would be a total nightmare, as you say.

Reply:
Taxons and shipping methods are probably not updated very often, don't use composition, and are connected to other objects in a fairly simple way where we can use `touch`, but most importantly: we fetch all taxons and all shipping methods in the database on every Darkswarm page.